src/placeos-driver/interface/sensor.cr

```diff
@@ -0,0 +1,137 @@
+require "json"
+require "s2_cells"
+
+module PlaceOS::Driver::Interface::Sensor
+  enum SensorType
+    Temperature # https://en.wikipedia.org/wiki/Celsius
+    Humidity    # https://en.wikipedia.org/wiki/Humidity#Relative_humidity (percentage)
+    Illuminance # https://en.wikipedia.org/wiki/Illuminance (lux)
+    Pressure    # https://en.wikipedia.org/wiki/Pascal_(unit)
+    Trigger     # Proximity or motion etc
+    Switch      # On or Off - water leak, door open etc
+    Level       # Percentage + raw value + technical human readable (battery, fuel, water tank, oxygen, Co2)
+    Flow        # https://en.wikipedia.org/wiki/Cubic_metre_per_second (gas or liquid)
+    Counter
+    Acceleration # MetrePerSecondSquared
+    Speed        # MetrePerSecond
+    # Gyroscope: roll, pitch, yaw - Angle
+    Roll
+    Pitch
+    Yaw
+    Compass              # direction (magnetic field)
+    Current              # https://en.wikipedia.org/wiki/Ampere
+    Voltage              # https://en.wikipedia.org/wiki/Volt
+    ElectricalResistance # https://en.wikipedia.org/wiki/Ohm
+    Radiation            # https://en.wikipedia.org/wiki/Sievert
+    Distance
+  end
+
+  # Using SI units and SI derived units
+  enum Unit
+    Celsius               # Temperature - celsius over Kelvin to avoid some conversions
+    Percentage            # Humidity, Level
+    Lux                   # Illuminance
+    Pascal                # Pressure
+    Boolean               # Trigger, Switch
+    Integer               # Counter
+    CubicMetrePerSecond   # Liquid or gas flow rate
+    MetrePerSecondSquared # Acceleration
+    MetrePerSecond        # Speed
+    Angle                 # Compass, Accel, Gyro
+    Ampere                # Current
+    Volt                  # Voltage
+    Ohm                   # ElectricalResistance
+    Sievert               # Radiation
+    Metre                 # Distance
+  end
+
+  # return the specified sensor details
+  abstract def sensor(unique_id : String) : Detail?
+
+  # return an array of sensor details
+  # zone_id can be ignored if location is unknown by the sensor provider
+  # mac_address can be used to grab data from a single device (basic grouping)
+  abstract def sensors(type : String? = nil, mac_address : String? = nil, zone_id : String? = nil) : Array(Detail)
+
+  abstract class Location
+    include JSON::Serializable
+
+    use_json_discriminator "type", {
+      "geo" => GeoLocation,
+      "map" => MapLocation,
+    }
+  end
+
+  class GeoLocation < Location
+    getter type : String = "geo"
+
+    def initialize(@lat, @lon)
+    end
+
+    property lat : Float64
+    property lon : Float64
+
+    def s2_cell_id(s2_level : Int32 = 21)
+      S2Cells::LatLon.new(lat, lon).to_token(s2_level)
+    end
+  end
```

Comment:
This may be cleaner to model as the s2 cell, with methods for then pulling lat-long if required.

S2 cells provide both a position and area, so dropping this to coords only results in info loss (that's potentially useful).

Reply:
The other one here (for both current location types) is altitude. Likely need to be able to resolve sensors to a floor within a building context. An driver implementing this interface may be providing sensors data across multiple floors so is not something that can be resolved by system zone membership.

Reply:
can add an optional altitude, these locations don't represent the location model we have at the area management level though, this is the raw data from the sensor platform - something we can use to generate an S2 cell, hence the helper method

Reply:
This same interface could be used in an abstraction though (or would be nice too)!

In that setup, there may set of point based sensors, which are then summarised and exposed as an area sensor.